mava/utils/training.py

```diff
@@ -56,3 +59,74 @@ def make_learning_rate(init_lr: float, config: DictConfig) -> Union[float, Calla
         return make_learning_rate_schedule(init_lr, config)
     else:
         return init_lr
+
+
+def select_action_ppo(
+    actor_output: Tuple[Array, Array],
+    key: PRNGKey,
+    env_name: str,
+) -> Tuple[Array, Array, Array]:
+    """Select action for the given actor output."""
+
+    actor_mean, actor_log_std = actor_output
+    actor_policy = distrax.MultivariateNormalDiag(actor_mean, jnp.exp(actor_log_std))
+
+    raw_action, log_prob = actor_policy.sample_and_log_prob(seed=key)
+    action, log_prob = transform_actions_log(env_name, raw_action, log_prob)
+
+    return raw_action, action, log_prob
+
+
+def get_logprob_entropy(
+    actor_output: Tuple[Array, Array],
+    traj_action: Array,
+    env_name: str,
+) -> Tuple[Array, Array]:
+    """Get the log probability and entropy of a given actor output and traj_batch action."""
+
+    actor_mean, actor_log_std = actor_output
+    actor_policy = distrax.MultivariateNormalDiag(actor_mean, jnp.exp(actor_log_std))
+
+    log_prob = actor_policy.log_prob(traj_action)
+    entropy = actor_policy.entropy().mean()
+
+    _, log_prob = transform_actions_log(env_name, traj_action, log_prob)
+    return log_prob, entropy
+
+
+def select_action_eval(
+    actor_output: Union[Tuple[Array, Array], Tuple[Array, Array, Array]],
+    key: PRNGKey,
+    env_name: str,
+    network: str = "feedforward",
+) -> Union[Array, Tuple[Array, Array]]:
+    """Select action for the given actor output."""
+    # This throws typing issues when committing:
+    # if network == "recurrent":
+    #     hstate, actor_mean, actor_log_std = actor_output
+    # else:
+    #     actor_mean, actor_log_std = actor_output
+
+    actor_policy = distrax.MultivariateNormalDiag(actor_output[-2], jnp.exp(actor_output[-1]))
+
+    raw_action = actor_policy.sample(seed=key)
+    action = transform_actions_log(env_name, raw_action, None)
+
+    return (actor_output[0], action) if network == "recurrent" else action
+
+
+def transform_actions_log(
+    env_name: str, raw_action: Array, log_prob: Array
+) -> Union[Array, Tuple[Array, Array]]:
+    """Transform action and log_prob values"""
+
+    # IF action in [-N, N] and N != 1 ELSE [-1, 1] and N == 1
+    n = 0.4 if env_name == "humanoid_9|8" else 1
+    action = n * jnp.tanh(raw_action)
+    if log_prob is None:
+        return action  # during evaluation.
+
+    # Note: jnp.log(derivative of action equation)
+    log_prob -= jnp.sum(jnp.log(n * (1 - jnp.tanh(raw_action) ** 2) + 1e-6), axis=-1)
```

Comment:
This might not be the best formulation of the transformation. I've been experimenting with 

```
jnp.sum(2 * (jnp.log(2) - unbound_action - jax.nn.softplus(-2 * unbound_action)), axis=-1)
```

Which might be a more numerically stable version. Can you test this and see if it works for you.